### DISCUSSION TITLE
When to use Python::gil and passing py in the method or function?

### DISCUSSION THREAD

**Discussion by Abhay2412** [PyO3/pyo3#4427] [Questions] [answered]
Hello, I have a question about the ```Python::with_gil``` in specific when should I be passing ```py``` to the function? For example, I have this function in Rust: 
```rs
#[pyfunction]
fn test(values: Vec<f32>, delay_value: usize) -> PyResult<()> {
    let training_file_path = Path::new("src\\assets\\white_500_2000.wav");
    let noise_data = read_wave_file(training_file_path).unwrap();
    test_actual(noise_data, values, delay);
    Ok(())
}
```
and then I add it to the module like this: 
```rs
#[pymodule]
fn python_rust_cancellation(m: &Bound<'_, PyModule>) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(test, m)?)?;
    Ok(())
}
``` 
So I am wondering if it is  required to pass in the ```py: Python``` to the ```test``` function? I tried reading the [documentation](https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#method.with_gil) on it regarding this it has to do with the global interpreter lock has to access the Python interpreter. If I have been doing this wrong it would be great to get some insight on this as I am new to Pyo3 and still not familiar with all the proper usage. This does work fine with me calling this Rust function in Python.

**Comment by birkenfeld** [comment, marked as answer]
It is not required to have `py: Python` as a function argument if you don't need it. The GIL will still be held and to release it you'll need to call `py.allow_threads()` (thereby requiring a `py` token!)

**Reply by Abhay2412** [reply]
Thank you for the swift reply. I appreciate it a lot for the clarification
